src/lsm/README.md

```diff
@@ -0,0 +1,111 @@
+    * Assert no compactions are currently running.
+    * Allow level visible table counts to overflow if needed.
+    * Start even level compactions if there's any tables to compact.
+
+2. (second) up beat of the bar:
+    * Finish ticking running even-level compactions.
+    * Assert on callback completion that all compactions are complete.
+
+3. (third) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Start odd level compactions if there are any tables to compact, and only if we must.
+    * Compact the immutable table if it contains any sorted values (it could be empty).
+
+4. (fourth) last beat of the bar:
+    * Finish ticking running odd-level and immutable table compactions.
+    * Assert on callback completion that all compactions are complete.
+    * Assert on callback completion that no level's visible table count overflows.
+    * Flush, clear, and sort mutable table values into immutable table for next bar.
+
+
+## Snapshots
+### Snapshots and Compaction
+
+Consider the half-bar compaction beginning at op=`X` (`12`), with `lsm_batch_multiple=M` (`8`).
+Each half-bar contains `N=M/2` (`4`) beats. The next half-bar begins at `Y=X+N` (`16`).
+
+During the compaction `X` (op=`X…Y-1`; `12…15`), each commit prefetches from the snapshot equal to
+the first op of the compaction. As shown, they continue to query the old (input) tables.
+
+During the compaction `X`:
+- `snapshot_max` of each input table is truncated to `Y-1` (`15`).
+- `snapshot_min` of each output table is initialized to `Y` (`16`).
+
+```
+  0   4   8  12  16  20  24  (op, snapshot)
+  ┼───┬───┼───┬───┼───┬───┼
+              ####
+  ····────────X────────····  (input  tables, before compaction)
+  ····────────────           (input  tables,  after compaction)
+                  Y────····  (output tables,  after compaction)
+```
+
+Beginning from the next op after the compaction (`Y`; `16`):
+- The output tables of the above compaction `X` are visible.
+- The input tables of the above compaction `X` are invisible.
+- Therefore, it will lookup from the output tables, but ignore the input tables.
+- Commits must not prefetch from the output tables of `X` before the compaction half-bar
+  has finished, since those tables are incomplete.
+
+At this point the input tables can be removed if they are invisible to all persistent snapshots.
+
+
+### Snapshot Values
+
+The on-disk tables visible to a snapshot `B` do not contain the updates from the commit with op `B`.
+
+Consider the following diagram (`lsm_batch_multiple=8`):
+
+```
+  0   4   8  12  16  20  24  28  (op, snapshot)
+  ┼───┬───┼───┬───┼───┬───┼───┬
+          ,,,,,,,,........
+          ↑A      ↑B      ↑C
+```
+
+Compaction is driven by the commits of ops `B→C` (`16…23`):
+- Updates from ops `A→B` (`8…15`) are in the immutable table.
+- Updates from ops `B→C` (`16…23`) are added to the mutable table (by the respective commit).
+- `lookup_snapshot_max` is `B` (`16`).
```

Comment:
What is this variable?

Reply:
Rewritten to:
```
- Lookups target snapshot `B` (`16`) (`Tree.lookup_snapshot_max`).
```

Reply:
Is this the snapshot against which op commits will do reads?

Reply:
Yes, commits always read against `tree.lookup_snapshot_max` (ignoring persistent snapshots).